util/winutil/s4u/s4u_windows.go

```diff
@@ -0,0 +1,804 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package s4u is an API for accessing Service-For-User (S4U) functionality on Windows.
+package s4u
+
+import (
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"os/user"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"tailscale.com/cmd/tailscaled/childproc"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/winutil"
+	"tailscale.com/util/winutil/conpty"
+)
+
+func init() {
+	childproc.Add("s4u", beRelay)
+}
+
+// TokenType is an enumeration specifying the type of S4U token to create.
+type TokenType uint
+
+const (
+	// TokenTypeQuery is an S4U token that may only be used for querying
+	// information about a user.
+	TokenTypeQuery TokenType = iota
```

Comment:
I originally named those values `Identification` and `Impersonation`, however as I fleshed out the API I realized that I didn't really want to expose a bunch of Windows NT concepts to the exported API. I'm imagining the scenario where somebody is working on a new feature (say, TailDrive) and wants to use S4U for something. I want these values to be meaningful and clear, but also understandable to somebody who isn't necessarily familiar with the concepts of identification vs impersonation.

I'm welcome to suggestions that could improve that clarity -- confusion is definitely something I want to avoid!

Reply:
I see your point, and I DO like the higher-level `s4u.Login` / `(*s4u.Session).Do` / `(*s4u.Session).StartProcess` set of API. I also expect it to be sufficient for most, if not all, scenarios. And we might want to provide higher-level APIs for some of the remaining scenarios.

For example, someone might need to access user's environment variables, or expand them in a given string. But in order for, say, `windows.Token.Environ` to work correctly and return user-specific variables, user's profile must be loaded. Additionally, and that's probably undocumented, some of the variables will only be set in the newly created environment block if the token is an impersonation (rather than identification) token. Therefore, I think it might be beneficial to allow s4u package users to access `s4u.Session`'s token and/or provide higher-level APIs that provide access to user's environment variables.

I do not oppose having the lower-level `s4u.CreateToken` API as exported. However I believe that its users must be familiar with the Windows NT concepts, and have a clear understanding of what it returns, and what it can and cannot be used for. From this perspective, I think it's better to make the names clear and searchable. And if we feel like we need to make it easier to use for non-Windows folks, maybe we could guide them in the documentation? Saying something like "An identification token can be used to query information about the user, while an impersonation token allows tailscaled to act on behalf of the user. CreateToken does not load the user's profile and cannot be used for operations like accessing the user's registry key or expanding user-specific environment variables. Consider using s4u.Login instead."

Reply:
That makes a lot of sense. I actually expose `CreateToken` because Tailscale SSH wants to query group memberships prior to logging in... Maybe the best way to deal with this is to make `CreateToken` internal and just expose the group-based API directly.